website/docs/plugin/framework/functions/errors.mdx

```diff
@@ -0,0 +1,251 @@
+---
+page_title: 'Plugin Development - Framework: Function Errors'
+description: |-
+  How to return function errors from the Terraform provider development
+  framework.
+---
+
+# Returning Errors
+
+Providers use `FunctionErrors` to surface errors generated during execution
+of provider-defined functions to practitioners. These errors are
+returned from Terraform CLI at the end of command output:
+
+```console
+$ terraform plan
+# ... other plan output ...
+╷
+│ Error: Message
+│ 
+│   on example.tf line #:
+│    #: source configuration line
+╵
+```
+
+In the framework, you may encounter them in response structs or as returns from
+provider-defined function execution.:
+
+```go
+func (f *ExampleFunction) Run(ctx context.Context, req function.RunRequest, resp *function.RunResponse) {
+```
+
+This is the most common form for Function Errors: a slice that has one or more
+errors appended to it. This approach allows your provider to inform
+practitioners about all relevant errors at the same time, allowing
+practitioners to fix their configuration or environment more quickly. You
+should only append to Function Errors slices and never replace or remove
+information from them.
+
+The next section will detail the concepts and typical behaviors of
+function errors, while the final section will outline the typical methods for
+working with function errors, using functionality from the available
+[`fwerror` package](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/fwerror).
+
+## Function Error Concepts
+
+### Error
+
+`Error` is a practitioner-oriented description of the problem. This should
+contain sufficient detail to provide a both general and more specific details
+regarding the issue. For example "Error executing function: foo can only contain
+letters, numbers, and digits."
+
+### Argument
+
+`Argument` identifies the specific function argument position that caused the
+error. Only errors that pertain to a function argument will
+include this information.
+
+## How Errors Affect State
+
+**Returning function errors does not stop the state from being updated**.
+Terraform will still persist the returned state even when function errors
+are returned with it. This is to allow Terraform to persist the values that have
+already been modified when a resource modification requires multiple API
+requests or an API request fails after an earlier one succeeded.
+
+When returning function errors, we recommend resetting the state in the
+response to the prior state available in the configuration.
```

Comment:
Can we double check this? I believe that typically function calls and their errors happen will so early in the operation graph that this would not necessarily be true (except maybe if there are unknown values being passed as arguments).

I'm wondering if its better to just omit this whole section, since "state" is not a function concept and this was mostly relating to managed resources, or whether we need to get into those function error handling implementation details.

Reply:
If it's ok, I'm going to hold off making any further changes to the documentation until we converge on the implementation of `FunctionError`/`FuncError`. But as an aside, I checked this, and you're quite right, returning a diagnostic will result in no alteration to the state file. Returning a function error currently has no effect, presumably because it is not yet being handled within core, and requires https://github.com/hashicorp/terraform/pull/34603 to be merged. Although, I believe I could build Terraform with this branch and test.

Reply:
I've updated the docs to remove this section, as suggested.